### DISCUSSION TITLE
Class decorators w/o strong typing: slight deviation from mypy

### DISCUSSION THREAD

**Discussion by michaelwooley** [microsoft/pyright#3268] [General]
Earlier today, I thought I'd run into a bonafide bug that arose when a class had a decorator attached to it. The original failure conditions were somewhat exotic (overloaded methods, complex inheritance, etc.). However, I eventually simplified my MWE to the point where I realized that the type signatures on the decorator were what actually mattered.

In the process of figuring that out, I noticed that mypy is slightly more forgiving if your decorator lacks full typing.

In the interest of feeling like I didn't lose the better part of a morning on this :man_facepalming:, I've pasted my resulting snippet below:

```python
# class_decorator_overload_q.py
from typing import Any, TypeVar  # , cast

TClass = TypeVar("TClass", bound=object) # Is bounding to `object` best for this case?


def decorated_poorly(cls: Any) -> Any:
    return cls


def decorated_correctly(cls: TClass) -> TClass:
    # return cast(TClass, cls) # Need this if return anything other than orig cls
    return cls


class Ex:
    def f(self, a: int) -> int:
        return 2 * a


@decorated_poorly
class ExAny:
    """Same as `Ex`, just have identity decorator on class."""

    def f(self, a: int) -> int:
        return 2 * a


@decorated_correctly
class ExTClass:
    """Same as `ExDecorator`, try spelling out hint in `run` method."""

    def f(self, a: int) -> int:
        return 2 * a


out1 = Ex().f(2)
reveal_type(out1)
out2 = ExAny().f(2)
reveal_type(out2)
out3 = ExTClass().f(2)
reveal_type(out3)
```

The results for the second case are different for mypy and pyright. In particular, mypy recognizes that the output of the second case is also an `int` while pyright says it is `Any`:

```bash
❯ pyright class_decorator_overload_q.py
No configuration file found.
No pyproject.toml file found.
Assuming Python platform Linux
Searching for source files
Found 1 source file
[....]/class_decorator_overload_q.py
  [....]/class_decorator_overload_q.py:38:13 - information: Type of "out1" is "int"
  [....]/class_decorator_overload_q.py:40:13 - information: Type of "out2" is "Any"
  [....]/class_decorator_overload_q.py:42:13 - information: Type of "out3" is "int"
0 errors, 0 warnings, 3 informations 
Completed in 0.509sec
❯ mypy class_decorator_overload_q.py
class_decorator_overload_q.py:38: note: Revealed type is "builtins.int"
class_decorator_overload_q.py:40: note: Revealed type is "builtins.int"
class_decorator_overload_q.py:42: note: Revealed type is "builtins.int"
```

Version info:

```bash
❯ python --version
Python 3.9.1
❯ mypy --version
mypy 0.910
❯ npm --version
8.5.4
❯ pyright --version
pyright 1.1.233
```

Thanks for the great package :)

**Comment by erictraut** [comment]
Thanks for the analysis. Yes, mypy appears to be incorrect in this case. If the decorator is explicitly annotated with a return type of `Any`, a type checker cannot make assumptions about the type. PEP 484 is quite clear on this matter. Mypy doesn't distinguish between an explicit `Any` and an implicit `Any`. Pyright does distinguish (and refers to the latter as `Unknown`.) Also, mypy doesn't ever attempt to infer return types of a function; it simply assumes a return type of `Any` if a function has no annotated return type. That may explain why mypy authors chose to deviate from PEP 484 rules in this case. 

You asked whether `TClass` should be bound to `object`. A TypeVar bound to `object` is effectively the same as a TypeVar with no bound. For this use case, I recommend the following instead:

```python
TClass = TypeVar("TClass")

def decorated_correctly(cls: type[TClass]) -> type[TClass]:
    return cls
```

**Reply by michaelwooley** [reply]
Thanks for your thorough reply, Eric. I'll do that going forward.
